### ISSUE TITLE
Passing ndarray.flat as an argument between njit compiled functions fails

### ISSUE THREAD

**Issue by lgarrison** [numba/numba#2954]
Accessing an array with `ndarray.flat` in a `numba.prange` loop causes an internal Numba error.  Here is a minimal example that demonstrates the exception:

```python
import numba
import numpy as np
import os
os.environ['NUMBA_DEBUG_ARRAY_OPT_STATS'] = '1'
        
@numba.njit(parallel=False)
def serial(A):
    A = A.flat
    N = len(A)
    for i in numba.prange(N):
        if A[i] >= 5:
            A[i] += 10
        
@numba.njit(parallel=True)
def parallel(A):
    A = A.flat
    N = len(A)
    for i in numba.prange(N):
        if A[i] >= 5:
            A[i] += 10

print('Serial:')
A = np.arange(10)
serial(A)
print(A)

print('\nParallel:')
A = np.arange(10)
parallel(A)
print(A)
```
and the output:
```
Serial:
[ 0  1  2  3  4 15 16 17 18 19]

Parallel:
Parallel for-loop #0 is produced from pattern '('prange', 'user')' at <ipython-input-1-f9ca7dd40116> (18)
After fusion, function parallel has 1 parallel for-loop(s) #set([0]).

---------------------------------------------------------------------------
LoweringError                             Traceback (most recent call last)
<ipython-input-1-f9ca7dd40116> in <module>()
     27 print('\nParallel:')
     28 A = np.arange(10)
---> 29 parallel(A)
     30 print(A)

/home/lgarrison/anaconda2/lib/python2.7/site-packages/numba/dispatcher.pyc in _compile_for_args(self, *args, **kws)
    358                     e.patch_message(''.join(e.args) + help_msg)
    359             # ignore the FULL_TRACEBACKS config, this needs reporting!
--> 360             raise e
    361 
    362     def inspect_llvm(self, signature=None):

LoweringError: Failed at nopython (nopython mode backend)
Failed at nopython (nopython mode backend)
expecting {{i8*, i8*, i64, i64, i64*, [1 x i64], [1 x i64]}, i64, i64*} but got i8*

File "<ipython-input-1-f9ca7dd40116>", line 18:
def parallel(A):
    <source elided>
    N = len(A)
    for i in numba.prange(N):
    ^
[1] During: lowering "id=0[LoopNest(index_variable = parfor_index.2, range = (0, $N.4, 1))]{40: <ir.Block at <ipython-input-1-f9ca7dd40116> (18)>, 81: <ir.Block at <ipython-input-1-f9ca7dd40116> (18)>, 59: <ir.Block at <ipython-input-1-f9ca7dd40116> (19)>}Var(parfor_index.2, <ipython-input-1-f9ca7dd40116> (18))" at <ipython-input-1-f9ca7dd40116> (18)
-------------------------------------------------------------------------------
This should not have happened, a problem has occurred in Numba's internals.

Please report the error message and traceback, along with a minimal reproducer
at: https://github.com/numba/numba/issues/new

If you need help writing a minimal reproducer please see:
http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports

If more help is needed please feel free to speak to the Numba core developers
directly at: https://gitter.im/numba/numba

Thanks in advance for your help in improving Numba!
```

The `if` statement appears necessary to cause the exception.  Without the `if`, the jit succeeds (issuing only a warning), but produces incorrect results (no modification to the input).

Keeping the `if` and using `ndarray.reshape(-1)` instead of `ndarray.flat` silently produces incorrect results (no warnings or errors issued).  So perhaps that's the most worrisome failure mode!  `ndarray.ravel()` looks okay in all cases.

**Comment by stuartarchibald** [comment]
Thanks for the report. I can reproduce. @DrTodd13 will take a look.

**Comment by DrTodd13** [comment]
@stuartarchibald Is the following expected to work?

```
@numba.njit(parallel=False)
def serial_internal(A):
    N = len(A)
    for i in numba.prange(N):
        if A[i] >= 5:
            A[i] += 10

@numba.njit(parallel=False)
def serial(A):
    serial_internal(A.flat)
```

For me, this non-ParallelAccelerator example seems to generate the same error that ParallelAccelerator has with the original code, which isn't surprising because ParallelAccelerator does this sort of function division.

There was a ParallelAccelerator bug as well that I at a minimum created a work-around for but still generated this same error.

**Comment by stuartarchibald** [comment]
@DrTodd13 looks like this is a bug for @njit not just ParallelAccelerator. Thanks for looking. Removing ParallelAccelerator label.

@lgarrison I'm editing the ticket title to reflect the discovered underlying issue.

Minimal reproducer:
```python
from numba import njit
import numpy as np

@njit
def bar(A):
    return A[0]

@njit
def foo(A):
    return bar(A.flat)

foo(np.arange(10))
```

gives this:
```
numba.errors.InternalError: Failed at nopython (nopython mode backend)
expecting {{i8*, i8*, i64, i64, i64*, [1 x i64], [1 x i64]}, i64, i64*} but got i8*
[1] During: resolving callee type: type(CPUDispatcher(<function bar at 0x7f102fd17ea0>))
[2] During: typing of call at foo.py (10)

--%<----------------------------------------------------------------------------


File "foo.py", line 10:
def foo(A):
    return bar(A.flat)
```

**Comment by ehsantn** [comment]
@stuartarchibald ParallelAccelerator does have an issue, which is alias analysis not understanding `A.flat` and is solved in #2963.

**Comment by stuartarchibald** [comment]
@ehsantn thanks, and thanks for the fix in #2963. I saw that go in and figured this can track the `@njit` issue which prevents both from working?

**Comment by ehsantn** [comment]
Sure.

**Comment by lgarrison** [comment]
Thanks!  Does this also fix the problem with `reshape(-1)`?  In my reproducer, if you replace `A.flat` with `A.reshape(-1)`, the input remains unmodified in the parallel case (but no errors).

**Comment by ehsantn** [comment]
@lgarrison Yes, #2963 will support `reshape` in general.

**Comment by lgarrison** [comment]
I don't think #2963 fixed this issue.  Running @stuartarchibald's reproducer gives me the same results.

I'm on Numba version 0.39.0dev0+139.gf9089c2dc (i.e. master).

**Comment by stuartarchibald** [comment]
@lgarrison thanks for checking back on this. The issue with Parallel Accelerator (alias analysis missing support for `ndarray.flat`) is fixed in #2963, but the underlying issue in the compiler core is not. To avoid confusion I've opened https://github.com/numba/numba/issues/3015 to track the specific compiler problem and I'm closing this ticket.
